### PR TITLE
feat(versioning): same major

### LINKED ISSUE #27312: New versioning: same-major
### Describe the proposed change(s).

We have a need in multiple areas to support the concept that a version like X.Y.Z means "Greater than or equal to X.Y.Z and less than X+1". e.g. "1.20.0" means the equivalent of "^1.20.0".

A recent example is https://github.com/renovatebot/renovate/discussions/27296#discussioncomment-8468506

I think we can create a simple versioning based off semver-coerced where every version is a constraint, and satisfies() returns true if the two versions are the same major version.

### PR DESCRIPTION
**Pull request by RahulGautamSingh**
## Changes
- Add a helper function `massageVersion` which massages single versions to constrains. eg. `X.Y.Z` as `>=X.Y.Z <X+1.0.0`

- Modify methods so that they can handle the version as a constraint and work as expected.
For eg. `matches` should return `true` for `matches('1.1.0', '1.0.0');`
For more examples, please see the tests.

## Context
Closes: https://github.com/renovatebot/renovate/issues/27312

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository
https://github.com/Rahul-renovate-testing/same-major

First run without same-major versioning created 2 PRs: 
i) for the highest version with same major as current version
ii) with the latest version

Second run with the same-major versioning  and PR i) was autoclosed, which is the expected behaviour.

### CODE REVIEW

File: lib/modules/versioning/same-major/index.spec.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
+import sameMajor from '.';
+
+describe('modules/versioning/same-major/index', () => {
+  describe('.matches(version, range)', () => {
+    it('should return true when version is has same major', () => {
```
Comment (RahulGautamSingh) [inline lib/modules/versioning/same-major/index.spec.ts:5]:
```suggestion
    it('should return true if the version has same major', () => {
```

File: lib/modules/versioning/same-major/index.spec.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+      expect(sameMajor.matches('1.0.1', '1.0.0')).toBeTrue();
+    });
+
+    it('should return false when version has different major', () => {
+      expect(sameMajor.matches('2.0.1', '1.0.0')).toBeFalse();
+    });
+
+    it('should return false when version is lesser than the current version', () => {
```
Comment (RahulGautamSingh) [inline lib/modules/versioning/same-major/index.spec.ts:13]:
```suggestion
    it('should return false when version is out of range', () => {
```
Add some more assertions

File: lib/modules/versioning/semver-coerced/index.spec.ts [hunk 3]
```diff
@@ -30,7 +30,7 @@ describe('modules/versioning/semver-coerced/index', () => {
     });
 
     it('invalid version', () => {
-      expect(semverCoerced.getMajor('xxx')).toBeNull();
+      expect(semverCoerced.getMajor('')).toBeNull();
```
Comment (viceice) [inline lib/modules/versioning/semver-coerced/index.spec.ts:33]: why this change?

File: lib/modules/versioning/same-major/index.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+ * Converts input to range if it's a version. eg. X.Y.Z -> '>=X.Y.Z <X+1.0.0'
+ * If the input is already a range, it returns the input.
+ */
+function massageVersion(input: string): string {
+  // return the input if it is a range
+  if (!semverCoerced.isSingleVersion(input)) {
+    return input;
+  }
+
+  // we are sure to get a majot because of the isSingleVersion check
```
Comment (viceice) [inline lib/modules/versioning/same-major/index.ts:20]:
```suggestion
  // we are sure to get a major because of the isSingleVersion check
```

File: lib/modules/versioning/same-major/index.ts [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+  const major = semverCoerced.getMajor(input)!;
+  return `>=${input} <${major + 1}.0.0`;
```
Comment (viceice) [inline lib/modules/versioning/same-major/index.ts:22]:
```suggestion
  return `>=${input} <${major + 1}`;
```
isn't this enough?

File: lib/modules/versioning/same-major/index.ts [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+function minSatisfyingVersion(
+  versions: string[],
+  range: string,
+): string | null {
+  return semverCoerced.minSatisfyingVersion(versions, massageVersion(range));
+}
+
+function isLessThanRange(version: string, range: string): boolean {
+  return !!semverCoerced.isLessThanRange?.(version, massageVersion(range));
```
Comment (viceice) [inline lib/modules/versioning/same-major/index.ts:44]:
```suggestion
  return semverCoerced.isLessThanRange!.(version, massageVersion(range));
```

We know it's there, so it should throw if not

**Review by viceice — COMMENTED** [review]
shouldn't we name it `semver-same-major` or `semver-coerced-same-major`?

Or at least the docs need to clarify that `semver-coerced` versioning is used under the hood

File: lib/modules/versioning/semver-coerced/index.spec.ts (hunk 3 above)
Comment (RahulGautamSingh) [reply lib/modules/versioning/semver-coerced/index.spec.ts]: was testing, forgot to revert.

File: lib/modules/versioning/same-major/index.ts (hunk 5 above)
Comment (RahulGautamSingh) [reply lib/modules/versioning/same-major/index.ts]: should be.

File: lib/modules/versioning/same-major/readme.md [hunk 7]
```diff
@@ -0,0 +1,6 @@
+Renovate's Same Major versioning is specifically designed to address scenarios where version specifications, denoted as X.Y.Z, signify a range of compatibility from greater than or equal to X.Y.Z to less than X+1.Y.Z. In essence, each individual version is to be treated as a constraint.
```
Comment (secustor) [inline lib/modules/versioning/same-major/readme.md:1]:
```suggestion
Renovate's 'Same Major' versioning is specifically designed to address scenarios where version specifications, denoted as X.Y.Z, signify a range of compatibility from greater than or equal to X.Y.Z to less than X+1.Y.Z. In essence, each individual version is to be treated as a constraint.
```

**Review by secustor — COMMENTED** [review]
Nitpick, but not blocking

**Review by secustor — DISMISSED** [review]

File: lib/modules/versioning/same-major/index.spec.ts [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+          ['1.0.0', '1.0.4', '1.3.0', '2.0.0'],
+          '1.0.3',
+        ),
+      ).toBe('1.0.4');
+    });
+  });
+
+  describe('.isLessThanRange(version, range)', () => {
+    it('should return true', () => {
+      expect(sameMajor.isLessThanRange?.('2.0.2', '3.0.0')).toBeTrue();
```
Comment (viceice) [inline lib/modules/versioning/same-major/index.spec.ts:49]:
```suggestion
      expect(sameMajor.isLessThanRange('2.0.2', '3.0.0')).toBeTrue();
```

File: lib/modules/versioning/same-major/index.spec.ts [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+    });
+
+    it('should return false', () => {
+      expect(sameMajor.isLessThanRange?.('4.0.0', '3.0.0')).toBeFalse();
+      expect(sameMajor.isLessThanRange?.('3.1.0', '3.0.0')).toBeFalse();
```
Comment (viceice) [inline lib/modules/versioning/same-major/index.spec.ts:54]:
```suggestion
      expect(sameMajor.isLessThanRange('4.0.0', '3.0.0')).toBeFalse();
      expect(sameMajor.isLessThanRange('3.1.0', '3.0.0')).toBeFalse();
```

File: lib/modules/versioning/same-major/index.ts [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+import type { VersioningApi } from '../types';
+
+export const id = 'same-major';
+export const displayName = 'Same Major Versioning';
+export const urls = [];
+export const supportsRanges = false;
+
+/**
+ *
+ * Converts input to range if it's a version. eg. X.Y.Z -> '>=X.Y.Z <X+1.0.0'
```
Comment (viceice) [inline lib/modules/versioning/same-major/index.ts:11]:
```suggestion
 * Converts input to range if it's a version. eg. X.Y.Z -> '>=X.Y.Z <X+1'
```

File: lib/modules/versioning/same-major/readme.md [hunk 11]
```diff
@@ -0,0 +1,6 @@
+Renovate's 'Same Major' versioning is specifically designed to address scenarios where version specifications, denoted as X.Y.Z, signify a range of compatibility from greater than or equal to X.Y.Z to less than X+1.Y.Z. In essence, each individual version is to be treated as a constraint.
```
Comment (viceice) [inline lib/modules/versioning/same-major/readme.md:1]:
```suggestion
Renovate's 'Same Major' versioning is specifically designed to address scenarios where version specifications, denoted as X.Y.Z, signify a range of compatibility from greater than or equal to X.Y.Z to less than X+1.
In essence, each individual version is to be treated as a constraint.
```

File: lib/modules/versioning/same-major/readme.md [hunk 12]
```diff
@@ -0,0 +1,6 @@
...
+
+This method is handy when managing dependencies like dotnet-sdk's rollForward settings. Let's say a project uses dotnet-sdk version 3.1.0. It needs to be compatible with any version in the 3.x.x range but not with versions in the next major version, like 4.x.x.
+
+This process employs Semver-Coerced versioning beneath the surface, wherein single versions (e.g., `X.Y.Z`) are converted to a range like
+`X+1.0.0` and then passed to the corresponding semver-coerced method.
```
Comment (viceice) [inline lib/modules/versioning/same-major/readme.md:6]:
```suggestion
This process employs Semver-Coerced versioning beneath the surface, wherein single versions (e.g., `X.Y.Z`) are converted to a range like `X+1` and then passed to the corresponding semver-coerced method.
```
one sentence per line

File: lib/modules/versioning/same-major/readme.md [hunk 13]
```diff
@@ -0,0 +1,6 @@
+Renovate's 'Same Major' versioning is specifically designed to address scenarios where version specifications, denoted as X.Y.Z, signify a range of compatibility from greater than or equal to X.Y.Z to less than X+1.Y.Z. In essence, each individual version is to be treated as a constraint.
+
+This method is handy when managing dependencies like dotnet-sdk's rollForward settings. Let's say a project uses dotnet-sdk version 3.1.0. It needs to be compatible with any version in the 3.x.x range but not with versions in the next major version, like 4.x.x.
```
Comment (viceice) [inline lib/modules/versioning/same-major/readme.md:3]:
```suggestion
This method is handy when managing dependencies like dotnet-sdk's rollForward settings.
Let's say a project uses dotnet-sdk version 3.1.0.
It needs to be compatible with any version in the 3.x.x range but not with versions in the next major version, like 4.x.x.
```

**Review by HonkingGoose — CHANGES_REQUESTED** [review]
The docs are hard to understand.

Please add some examples that explain how it works. Like maybe a diff of "old dependency" vs "proposed Renovate update".

Also focus on the reader, what problem is this feature fixing, how to use/configure it, and how to know if it's working as intended.

File: lib/modules/versioning/same-major/readme.md [hunk 14]
```diff
@@ -0,0 +1,8 @@
+Renovate's 'Same Major' versioning is specifically designed to address scenarios where version specifications, denoted as X.Y.Z, signify a range of compatibility from greater than or equal to X.Y.Z to less than X+1.
```
Comment (HonkingGoose) [inline lib/modules/versioning/same-major/readme.md:1]:
```suggestion
Renovate's 'Same Major' versioning is designed to address scenarios where version specifications, denoted as `X.Y.Z`, signify a range of compatibility from greater than or equal to `X.Y.Z` to less than `X+1`.
```

File: lib/modules/versioning/same-major/readme.md [hunk 15]
```diff
@@ -0,0 +1,8 @@
...
+In essence, each individual version is to be treated as a constraint.
+
+This method is handy when managing dependencies like dotnet-sdk's rollForward settings.
+Let's say a project uses dotnet-sdk version 3.1.0.
```
Comment (HonkingGoose) [inline lib/modules/versioning/same-major/readme.md:5]:
```suggestion
Let's say a project uses dotnet-sdk version `3.1.0`.
```

File: lib/modules/versioning/same-major/readme.md [hunk 16]
```diff
@@ -0,0 +1,8 @@
...
+It needs to be compatible with any version in the 3.x.x range but not with versions in the next major version, like 4.x.x.
```
Comment (HonkingGoose) [inline lib/modules/versioning/same-major/readme.md:6]:
```suggestion
It needs to be compatible with any version in the `3.x.x` range but _not_ with versions in the next major version, like `4.x.x`.
```

File: lib/modules/versioning/same-major/readme.md [hunk 17]
```diff
@@ -0,0 +1,8 @@
...
+
+This process employs Semver-Coerced versioning beneath the surface, wherein single versions (e.g., `X.Y.Z`) are converted to a range like `X+1` and then passed to the corresponding semver-coerced method.
```
Comment (HonkingGoose) [inline lib/modules/versioning/same-major/readme.md:8]:
```suggestion
This process uses Semver-Coerced versioning beneath the surface, single versions (e.g., `X.Y.Z`) are converted to a range like `X+1` and then passed to the corresponding semver-coerced method.
```

**Review by HonkingGoose — CHANGES_REQUESTED** [review]
Some style fixes for the current text. I still find this hard to read and understand.

File: lib/modules/versioning/same-major/index.ts [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+
+  // we are sure to get a major because of the isSingleVersion check
+  const major = semverCoerced.getMajor(input)!;
+  return `>=${input} <${major + 1}`;
+}
+
+// for same major versioning one version is greater than the other if its major is greater
+function isGreaterThan(version: string, other: string): boolean {
+  const versionMajor = semverCoerced.getMajor(version)!;
+  const otherMajor = semverCoerced.getMajor(other)!;
```
[the remark this answers is not in the document]
Comment (RahulGautamSingh) [reply lib/modules/versioning/same-major/index.ts:28]: The name's same major so we will only use it for versions which have a major.

[the remark this answers is not in the document]
Comment (secustor) [reply lib/modules/versioning/same-major/index.ts:28]: Though users could supply unparseable strings, then this will return null and therefore lead to unexpected behaviour.

File: lib/modules/versioning/same-major/readme.md [hunk 19]
```diff
@@ -0,0 +1,28 @@
+The 'Same Major' versioning is a `unique` type of versioning that Renovate provides, which is not implemented t handle a manager or datasource.
+But instead to provide a way for users to handle the case where a version needs to treated as a constraint.
```
Comment (rarkins) [inline lib/modules/versioning/same-major/readme.md:2]:
```suggestion
The 'Same Major' versioning is designed to handle the case where a version needs to treated as a "greate than or equal to" constraint.
```

File: lib/modules/versioning/same-major/index.spec.ts [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+      expect(sameMajor.isGreaterThan('3.1.0', '3.0.0')).toBeFalse(); // same major -> equal
+      expect(sameMajor.isGreaterThan('3.0.0', '3.0.0')).toBeFalse(); // equal
+      expect(sameMajor.isGreaterThan('a', '3.0.0')).toBeFalse(); // invalid versions
+    });
+  });
+
+  describe('.matches(version, range)', () => {
+    it('should return true when version has same major', () => {
+      expect(sameMajor.matches('1.0.1', '1.0.0')).toBeTrue();
+      expect(sameMajor.matches('1.0.1', '^1.0.0')).toBeTrue(); // coverage for when ranges are passed to the function
```
Comment (rarkins) [inline lib/modules/versioning/same-major/index.spec.ts:20]: Why are we supporting range syntax like this?

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/modules/versioning/same-major/index.spec.ts (hunk 20 above)
Comment (RahulGautamSingh) [reply lib/modules/versioning/same-major/index.spec.ts:20]: We return the version as is without massaging if it's a range, which leads to this behaviour. Should I throw an error instead or log a warning instead reminding the user to not use same-major with ranges.

Comment (rarkins) [reply lib/modules/versioning/same-major/index.spec.ts:20]: I'm ok with "accidentally" supporting more than we want (e.g. because semver-coerced supports it) but we shouldn't test or document it as that makes it semi-official

**Review by rarkins — APPROVED** [review]
